### ISSUE TITLE
2D Heat Conduction for a Composite Geometry Not Working

### ISSUE THREAD

**Issue by alokwarey** [lululxvi/deepxde#231]
Trying to solve a 2D heat conduction problem with a circle inside a square. The BC's seem to be set correctly, but the solution is completely off. Not sure what I am doing wrong. 2D heat conduction without the circle in the center works. Any help/guidance is appreciated!

def pde(x, T):
dT_x = tf.gradients(T, x)[0]
dT_x, dT_y = dT_x[:, 0:1], dT_x[:, 1:2]
dT_xx = tf.gradients(dT_x, x)[0][:, 0:1]
dT_yy = tf.gradients(dT_y, x)[0][:, 1:2]
return dT_xx + dT_yy # Heat conduction equation

domain = dde.geometry.Rectangle([0,0],[1,1])
circle= dde.geometry.Disk(center= [0.5,0.5], radius= 0.1)

geom = dde.geometry.CSGDifference(domain, circle)

def boundary_x_l(x, on_boundary):
return on_boundary and np.isclose(x[0], 0)

def boundary_x_r(x, on_boundary):
return on_boundary and np.isclose(x[0], 1)

def boundary_y_b(x, on_boundary):
return on_boundary and np.isclose(x[1], 0)

def boundary_y_u(x, on_boundary):
return on_boundary and np.isclose(x[1], 1)

**Tried the fowllowing two options:**
OPTION 1
def boundary_circle(x, on_boundary):
return on_boundary and circle.on_boundary(x[:2])

OPTION 2
def boundary_circle(x, on_boundary):
return on_boundary and np.isclose((x[0]- 0.5)^2 + (x[1]-0.5)*^2 - 0.1^2, 0.0)

def func_0(x):
return np.zeros((len(x),1), dtype=np.float32)

def func_1(x):
return np.ones((len(x),1), dtype=np.float32)

bc_x_l = dde.DirichletBC(geom, func_0, boundary_x_l)
bc_x_r = dde.DirichletBC(geom, func_0, boundary_x_r)
bc_y_u = dde.DirichletBC(geom, func_0, boundary_y_u)
bc_y_b = dde.DirichletBC(geom, func_0, boundary_y_b)
bc_circle = dde.DirichletBC(geom, func_1, boundary_circle)

data = dde.data.PDE(
geom, pde, [bc_x_l, bc_x_r, bc_y_b, bc_y_u, bc_circle], num_domain=5000, num_boundary=1000,
num_test=None)

net = dde.maps.FNN([2] + [20] * 8 + [1], "tanh", "Glorot normal")

model = dde.Model(data, net)

model.compile("adam", lr=1e-3)
model.train(epochs=5000)
model.compile("L-BFGS-B")
losshistory, train_state = model.train()
dde.saveplot(losshistory, train_state, issave=True, isplot=True)

![image](https://user-images.githubusercontent.com/69644995/108546200-cb93af80-72b6-11eb-964c-c7e006687cfc.png)

![image](https://user-images.githubusercontent.com/69644995/108546233-d5b5ae00-72b6-11eb-8219-ee98568f732e.png)

![image](https://user-images.githubusercontent.com/69644995/108546274-e82fe780-72b6-11eb-8c17-81c60e9ef756.png)

**Comment by smao-astro** [comment]
Hi @alokwarey , I have tried your example, looks weird, could you also post the analytical/numerical solution here, thanks.

**Comment by smao-astro** [comment]
I can make it converge if set both outer and inter bc to zeros or both to ones.

**Comment by alokwarey** [comment]
@smao-astro : Yes I can solve the simple 2D heat conduction problem as well. The problem is with the composite geometry and different BC's. Here is the correct solution:

![Untitled](https://user-images.githubusercontent.com/69644995/108602517-4a98ee80-7370-11eb-8cd2-c69b2f5b97d2.png)

**Comment by alokwarey** [comment]
@smao-astro: I got it to work.

Had to assign more weights to the BC loss than PDE loss:

net = dde.maps.FNN([2] + [50] * 4 + [1], "tanh", "Glorot uniform")

model = dde.Model(data, net)

model.compile("adam", lr=1e-3, loss_weights = [1, 50, 50, 50, 50, 50])
model.train(epochs=15000)
model.compile("L-BFGS-B")
losshistory, train_state = model.train()
dde.saveplot(losshistory, train_state, issave=True, isplot=True)

![image](https://user-images.githubusercontent.com/69644995/108609006-d32b8500-7398-11eb-9f24-5dace325b3cf.png)
